mlir/lib/Quantum/Transforms/PropagateSimpleStatesAnalysis.hpp

```diff
@@ -0,0 +1,236 @@
+// Copyright 2024 Xanadu Quantum Technologies Inc.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// This algorithm is taken from https://arxiv.org/pdf/2012.07711, figure 5
+
+#pragma once
+
+#include <map>
+
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/Pass/Pass.h"
+#include "llvm/ADT/DenseMap.h"
+#include "llvm/Support/Debug.h"
+
+#include "Catalyst/IR/CatalystDialect.h"
+#include "Quantum/IR/QuantumOps.h"
+
+using namespace llvm;
+using namespace mlir;
+using namespace catalyst;
+
+namespace catalyst {
+
+// The six Pauli eigenstates
+enum class QubitState {
+    ZERO,
+    ONE,
+    PLUS,
+    MINUS,
+    LEFT,
+    RIGHT,
+    NOT_A_BASIS,
+};
+
+// {input state : {gate, output state}}
+static std::map<QubitState, std::map<StringRef, QubitState>> QubitTransitions = {
```

Comment:
This FSM ignores global phases. Any potential pitfalls?

Reply:
Optimizations should be valid regardless and typically do not consider global phases anyways. In some circumstances it is nice if the simulation respects the global phase, but I don't think anyone would have this expectation if they want the compiler to optimize :)